engine-types/src/types.rs

```diff
@@ -104,6 +108,114 @@ impl Mul<EthGas> for u64 {
     }
 }
 
+#[derive(
+    Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, BorshSerialize, BorshDeserialize,
+)]
+/// Engine `fee` type which wraps an underlying u128.
+pub struct Fee(u128);
+
+impl Display for Fee {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Fee {
+    /// Constructs a new `Fee` with a given u128 value.
+    pub const fn new(fee: u128) -> Fee {
+        Self(fee)
+    }
+
+    /// Consumes `Fee` and returns the underlying type.
+    pub fn into_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl Add<Fee> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 + rhs.0)
+    }
+}
+
+impl Add<Fee> for u128 {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self + rhs.0)
+    }
+}
+
+impl Add<u128> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: u128) -> Self::Output {
+        Fee(self.0 + rhs)
+    }
+}
+
+impl Mul<Fee> for Fee {
+    type Output = Fee;
+
+    fn mul(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 * rhs.0)
+    }
+}
```

Comment:
I don't think this implementation should exist. A fee times a fee is not a fee, it's some kind of "fee area" (or similar nonsense concept).

When you're defining a new type you are giving semantics to a type; it's not a pure number anymore. Imagine if you were defining a `Meter` new type. This `impl Mul<Meter> for Meter` would not output `Meter`, it would output `SquareMeter` because areas are semantically different from lengths.

The reason it is important to not have implementations which break the semantics of the type is because otherwise it would be possible to write buggy logic. If you find yourself needing to do a `Fee * Fee` operation then the compiler should flag to you that it is not allowed because it means either you have assigned the types wrong (one of the arguments is actually a pure number, not a fee) or something is logically incorrect in the computation.

Note it is ok to have `impl Mul<u128> for Fee` because `u128` is still a pure number. This implementation is effectively a "scalar multiplication" (in mathematics jargon), or you could think about it as a short-cut for repeated addition (and of course `impl Add<Fee> for Fee` should output another `Fee`).

Reply:
The fact is, we are not using multiplication at this time.

Otherwise, I disagree. A simple example - we can do Watts.
2 watts * 3 watts = 6 watts.

This means operating with a single entity, the result of which is the same entity.

Reply:
Was removed

Reply:
Thanks for removing.

Watts times Watts is not Watts; it would be Watts^2, which is not a unit of power. When you multiply quantities you get a quantity which has the units of both inputs. When you add quantities, you and only add ones with the same units. This is an application of [dimensional analysis](https://en.wikipedia.org/wiki/Dimensional_analysis).